{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e976c838-6dcb-45ea-a743-1d70707ce535",
   "metadata": {},
   "source": [
    "# Chapter 15: Macromolecular Structures\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "56662937-d1e0-4246-9da4-dab815fc8486",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# function for Obtaining structure data\n",
    "\n",
    "from urllib.request import urlopen\n",
    "\n",
    "# %s indicating where the database identifier will be inserted\n",
    "#PDB_URL = 'http://files.rcsb.org/header/%s.pdb?format=PDB&compression=None'\n",
    "PDB_URL = 'https://files.rcsb.org/download/%s.pdb'\n",
    "# function to download pdb file \n",
    "def downloadPDB(pdbId, fileName=None):\n",
    "    \n",
    "    if not fileName:\n",
    "        fileName = '%s.pdb' % pdbId\n",
    "        \n",
    "    response = urlopen(PDB_URL % pdbId)\n",
    "    data = response.read().decode('utf-8')\n",
    "    \n",
    "    fileObj = open(fileName, 'w')\n",
    "    fileObj.write(data)\n",
    "    fileObj.close()\n",
    "    \n",
    "    return fileName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "da596116-4609-4ea4-ac02-4ecf1c5ef87f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'audit_author': [{'name': 'Tame, J.', 'pdbx_ordinal': 1},\n",
       "  {'name': 'Vallone, B.', 'pdbx_ordinal': 2}],\n",
       " 'cell': {'angle_alpha': 90.0,\n",
       "  'angle_beta': 99.61,\n",
       "  'angle_gamma': 90.0,\n",
       "  'length_a': 62.65,\n",
       "  'length_b': 82.43,\n",
       "  'length_c': 53.53,\n",
       "  'zpdb': 4},\n",
       " 'citation': [{'country': 'DK',\n",
       "   'id': 'primary',\n",
       "   'journal_abbrev': 'Acta Crystallogr.,Sect.D',\n",
       "   'journal_id_astm': 'ABCRE6',\n",
       "   'journal_id_csd': '0766',\n",
       "   'journal_id_issn': '0907-4449',\n",
       "   'journal_volume': '56',\n",
       "   'page_first': '805',\n",
       "   'page_last': '811',\n",
       "   'pdbx_database_id_doi': '10.1107/S0907444900006387',\n",
       "   'pdbx_database_id_pub_med': 10930827,\n",
       "   'rcsb_authors': ['Tame, J.R.', 'Vallone, B.'],\n",
       "   'rcsb_is_primary': 'Y',\n",
       "   'rcsb_journal_abbrev': 'Acta Crystallogr D Biol Crystallogr',\n",
       "   'title': 'The structures of deoxy human haemoglobin and the mutant Hb Tyralpha42His at 120 K.',\n",
       "   'year': 2000}],\n",
       " 'diffrn': [{'ambient_temp': 120.0, 'crystal_id': '1', 'id': '1'}],\n",
       " 'diffrn_detector': [{'details': 'MIRROR',\n",
       "   'detector': 'IMAGE PLATE',\n",
       "   'diffrn_id': '1',\n",
       "   'pdbx_collection_date': '1997-05-20T00:00:00+0000',\n",
       "   'type': 'RIGAKU RAXIS II'}],\n",
       " 'diffrn_radiation': [{'diffrn_id': '1',\n",
       "   'monochromator': 'NI FILTER',\n",
       "   'pdbx_monochromatic_or_laue_ml': 'M',\n",
       "   'pdbx_scattering_type': 'x-ray',\n",
       "   'wavelength_id': '1'}],\n",
       " 'diffrn_source': [{'diffrn_id': '1', 'pdbx_wavelength': '1.5418'}],\n",
       " 'entry': {'id': '1A3N'},\n",
       " 'exptl': [{'crystals_number': 1, 'method': 'X-RAY DIFFRACTION'}],\n",
       " 'exptl_crystal': [{'density_matthews': 2.19,\n",
       "   'density_percent_sol': 43.94,\n",
       "   'id': '1'}],\n",
       " 'pdbx_audit_revision_details': [{'data_content_type': 'Structure model',\n",
       "   'ordinal': 1,\n",
       "   'provider': 'repository',\n",
       "   'revision_ordinal': 1,\n",
       "   'type': 'Initial release'}],\n",
       " 'pdbx_audit_revision_group': [{'data_content_type': 'Structure model',\n",
       "   'group': 'Version format compliance',\n",
       "   'ordinal': 1,\n",
       "   'revision_ordinal': 2},\n",
       "  {'data_content_type': 'Structure model',\n",
       "   'group': 'Version format compliance',\n",
       "   'ordinal': 2,\n",
       "   'revision_ordinal': 3}],\n",
       " 'pdbx_audit_revision_history': [{'data_content_type': 'Structure model',\n",
       "   'major_revision': 1,\n",
       "   'minor_revision': 0,\n",
       "   'ordinal': 1,\n",
       "   'revision_date': '1998-04-29T00:00:00+0000'},\n",
       "  {'data_content_type': 'Structure model',\n",
       "   'major_revision': 1,\n",
       "   'minor_revision': 1,\n",
       "   'ordinal': 2,\n",
       "   'revision_date': '2008-03-24T00:00:00+0000'},\n",
       "  {'data_content_type': 'Structure model',\n",
       "   'major_revision': 1,\n",
       "   'minor_revision': 2,\n",
       "   'ordinal': 3,\n",
       "   'revision_date': '2011-07-13T00:00:00+0000'}],\n",
       " 'pdbx_database_status': {'pdb_format_compatible': 'Y',\n",
       "  'recvd_initial_deposition_date': '1998-01-22T00:00:00+0000',\n",
       "  'status_code': 'REL',\n",
       "  'status_code_sf': 'REL'},\n",
       " 'pdbx_vrpt_summary': {'babinet_b': 58.721,\n",
       "  'babinet_k': 0.202,\n",
       "  'fo_fc_correlation': 0.958,\n",
       "  'wilson_baniso': '[12.714,19.263,12.049,0.000,-1.530,0.000]',\n",
       "  'wilson_bestimate': 14.514,\n",
       "  'absolute_percentile_dccrfree': 77.0,\n",
       "  'absolute_percentile_clashscore': 87.7,\n",
       "  'absolute_percentile_percent_rsrzoutliers': 90.6,\n",
       "  'absolute_percentile_percent_ramachandran_outliers': 100.0,\n",
       "  'absolute_percentile_percent_rotamer_outliers': 51.5,\n",
       "  'acentric_outliers': 2,\n",
       "  'angles_rmsz': 1.31,\n",
       "  'attempted_validation_steps': 'molprobity,validation-pack,xtriage,eds,mogul,buster-report,percentiles,writexml',\n",
       "  'bonds_rmsz': 0.84,\n",
       "  'bulk_solvent_b': 55.889,\n",
       "  'bulk_solvent_k': 0.399,\n",
       "  'ccp4version': '7.0.044 (Gargrove)',\n",
       "  'centric_outliers': 0.0,\n",
       "  'clashscore': 2.02,\n",
       "  'data_anisotropy': 0.492,\n",
       "  'data_completeness': 95.43,\n",
       "  'high_resol_relative_percentile_dccrfree': 1.8,\n",
       "  'high_resol_relative_percentile_clashscore': 1.8,\n",
       "  'high_resol_relative_percentile_percent_rsrzoutliers': 1.8,\n",
       "  'high_resol_relative_percentile_percent_ramachandran_outliers': 1.8,\n",
       "  'high_resol_relative_percentile_percent_rotamer_outliers': 1.8,\n",
       "  'ligands_for_buster_report': 'yes',\n",
       "  'low_resol_relative_percentile_dccrfree': 1.8,\n",
       "  'low_resol_relative_percentile_clashscore': 1.8,\n",
       "  'low_resol_relative_percentile_percent_rsrzoutliers': 1.8,\n",
       "  'low_resol_relative_percentile_percent_ramachandran_outliers': 1.8,\n",
       "  'low_resol_relative_percentile_percent_rotamer_outliers': 1.8,\n",
       "  'num_hreduce': 4478.0,\n",
       "  'num_pdbids_absolute_percentile_dccrfree': 130704,\n",
       "  'num_pdbids_absolute_percentile_clashscore': 141614,\n",
       "  'num_pdbids_absolute_percentile_percent_rsrzoutliers': 127900,\n",
       "  'num_pdbids_absolute_percentile_percent_ramachandran_outliers': 138981,\n",
       "  'num_pdbids_absolute_percentile_percent_rotamer_outliers': 138945,\n",
       "  'num_pdbids_relative_percentile_dccrfree': 5950,\n",
       "  'num_pdbids_relative_percentile_clashscore': 6793,\n",
       "  'num_pdbids_relative_percentile_percent_rsrzoutliers': 5850,\n",
       "  'num_pdbids_relative_percentile_percent_ramachandran_outliers': 6697,\n",
       "  'num_pdbids_relative_percentile_percent_rotamer_outliers': 6696,\n",
       "  'num_angles_rmsz': 6094,\n",
       "  'num_bonds_rmsz': 4486,\n",
       "  'num_free_reflections': 2389,\n",
       "  'num_miller_indices': 47376,\n",
       "  'percent_rsrzoutliers': 0.52,\n",
       "  'percent_free_reflections': 5.04,\n",
       "  'percent_ramachandran_outliers': 0.0,\n",
       "  'percent_rotamer_outliers': 2.17,\n",
       "  'percentilebins': ['all', '1.8', 'xray'],\n",
       "  'protein_dnarnaentities': '1,2',\n",
       "  'refmac_version': '5.8.0158',\n",
       "  'relative_percentile_dccrfree': 70.9,\n",
       "  'relative_percentile_clashscore': 86.2,\n",
       "  'relative_percentile_percent_rsrzoutliers': 88.9,\n",
       "  'relative_percentile_percent_ramachandran_outliers': 100.0,\n",
       "  'relative_percentile_percent_rotamer_outliers': 39.4,\n",
       "  'report_creation_date': 'May 13, 2020 -- 02:42 pm BST',\n",
       "  'restypes_notchecked_for_bond_angle_geometry': ['HEM'],\n",
       "  'trans_nsc': 'The largest off-origin peak in the Patterson function is 6.92% of the height of the origin peak. No significant pseudotranslation is detected.',\n",
       "  'twin_l': 0.493,\n",
       "  'twin_l2': 0.325,\n",
       "  'xtriage_input_columns': 'F_meas_au,F_meas_sigma_au',\n",
       "  'bfactor_type': 'FULL',\n",
       "  'dccr': 0.16,\n",
       "  'dccrfree': 0.21,\n",
       "  'dccrefinement_program': 'REFMAC',\n",
       "  'edsr': 0.16,\n",
       "  'edsresolution': 1.8,\n",
       "  'edsresolution_low': 14.9,\n",
       "  'iover_sigma': '1.29(1.79A)',\n",
       "  'pdbr': 0.17,\n",
       "  'pdbrfree': 0.22,\n",
       "  'pdbdeposition_date': '1998-01-22T00:00:00+0000',\n",
       "  'pdbresolution': 1.8,\n",
       "  'pdbresolution_low': 15.0,\n",
       "  'pdbrevision_date': '2011-07-13T00:00:00+0000',\n",
       "  'pdbrevision_number': 3.0},\n",
       " 'rcsb_accession_info': {'deposit_date': '1998-01-22T00:00:00+0000',\n",
       "  'has_released_experimental_data': 'Y',\n",
       "  'initial_release_date': '1998-04-29T00:00:00+0000',\n",
       "  'major_revision': 1,\n",
       "  'minor_revision': 2,\n",
       "  'revision_date': '2011-07-13T00:00:00+0000',\n",
       "  'status_code': 'REL'},\n",
       " 'rcsb_entry_container_identifiers': {'assembly_ids': ['1'],\n",
       "  'entity_ids': ['1', '2', '3', '4'],\n",
       "  'entry_id': '1A3N',\n",
       "  'model_ids': [1],\n",
       "  'non_polymer_entity_ids': ['3'],\n",
       "  'polymer_entity_ids': ['1', '2'],\n",
       "  'rcsb_id': '1A3N',\n",
       "  'pubmed_id': 10930827},\n",
       " 'rcsb_entry_info': {'assembly_count': 1,\n",
       "  'branched_entity_count': 0,\n",
       "  'cis_peptide_count': 0,\n",
       "  'deposited_atom_count': 4997,\n",
       "  'deposited_hydrogen_atom_count': 0,\n",
       "  'deposited_model_count': 1,\n",
       "  'deposited_modeled_polymer_monomer_count': 572,\n",
       "  'deposited_nonpolymer_entity_instance_count': 4,\n",
       "  'deposited_polymer_entity_instance_count': 4,\n",
       "  'deposited_polymer_monomer_count': 574,\n",
       "  'deposited_solvent_atom_count': 455,\n",
       "  'deposited_unmodeled_polymer_monomer_count': 2,\n",
       "  'diffrn_radiation_wavelength_maximum': 1.5418,\n",
       "  'diffrn_radiation_wavelength_minimum': 1.5418,\n",
       "  'disulfide_bond_count': 0,\n",
       "  'entity_count': 4,\n",
       "  'experimental_method': 'X-ray',\n",
       "  'experimental_method_count': 1,\n",
       "  'inter_mol_covalent_bond_count': 0,\n",
       "  'inter_mol_metalic_bond_count': 4,\n",
       "  'molecular_weight': 64.55,\n",
       "  'na_polymer_entity_types': 'Other',\n",
       "  'nonpolymer_bound_components': ['HEM'],\n",
       "  'nonpolymer_entity_count': 1,\n",
       "  'nonpolymer_molecular_weight_maximum': 0.62,\n",
       "  'nonpolymer_molecular_weight_minimum': 0.62,\n",
       "  'polymer_composition': 'heteromeric protein',\n",
       "  'polymer_entity_count': 2,\n",
       "  'polymer_entity_count_dna': 0,\n",
       "  'polymer_entity_count_rna': 0,\n",
       "  'polymer_entity_count_nucleic_acid': 0,\n",
       "  'polymer_entity_count_nucleic_acid_hybrid': 0,\n",
       "  'polymer_entity_count_protein': 2,\n",
       "  'polymer_entity_taxonomy_count': 2,\n",
       "  'polymer_molecular_weight_maximum': 15.89,\n",
       "  'polymer_molecular_weight_minimum': 15.15,\n",
       "  'polymer_monomer_count_maximum': 146,\n",
       "  'polymer_monomer_count_minimum': 141,\n",
       "  'resolution_combined': [1.8],\n",
       "  'selected_polymer_entity_types': 'Protein (only)',\n",
       "  'software_programs_combined': ['DENZO', 'REFMAC', 'SCALEPACK'],\n",
       "  'solvent_entity_count': 1,\n",
       "  'diffrn_resolution_high': {'provenance_source': 'Depositor assigned',\n",
       "   'value': 1.8}},\n",
       " 'rcsb_primary_citation': {'country': 'DK',\n",
       "  'id': 'primary',\n",
       "  'journal_abbrev': 'Acta Crystallogr.,Sect.D',\n",
       "  'journal_id_astm': 'ABCRE6',\n",
       "  'journal_id_csd': '0766',\n",
       "  'journal_id_issn': '0907-4449',\n",
       "  'journal_volume': '56',\n",
       "  'page_first': '805',\n",
       "  'page_last': '811',\n",
       "  'pdbx_database_id_doi': '10.1107/S0907444900006387',\n",
       "  'pdbx_database_id_pub_med': 10930827,\n",
       "  'rcsb_authors': ['Tame, J.R.', 'Vallone, B.'],\n",
       "  'rcsb_journal_abbrev': 'Acta Crystallogr D Biol Crystallogr',\n",
       "  'title': 'The structures of deoxy human haemoglobin and the mutant Hb Tyralpha42His at 120 K.',\n",
       "  'year': 2000},\n",
       " 'refine': [{'ls_rfactor_rfree': 0.22038,\n",
       "   'ls_rfactor_rwork': 0.17076,\n",
       "   'ls_dres_high': 1.8,\n",
       "   'ls_dres_low': 15.0,\n",
       "   'ls_number_reflns_rfree': 2000,\n",
       "   'ls_number_reflns_obs': 49661,\n",
       "   'ls_percent_reflns_rfree': 5.0,\n",
       "   'ls_percent_reflns_obs': 95.78,\n",
       "   'pdbx_diffrn_id': ['1'],\n",
       "   'pdbx_ls_sigma_f': 0.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'biso_mean': 16.2}],\n",
       " 'refine_hist': [{'cycle_id': 'LAST',\n",
       "   'number_atoms_solvent': 455,\n",
       "   'number_atoms_total': 4997,\n",
       "   'pdbx_number_atoms_ligand': 172,\n",
       "   'pdbx_number_atoms_nucleic_acid': 0,\n",
       "   'pdbx_number_atoms_protein': 4370,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'dres_high': 1.8,\n",
       "   'dres_low': 15.0}],\n",
       " 'refine_ls_restr': [{'dev_ideal': 0.012,\n",
       "   'dev_ideal_target': 0.02,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_bond_d'},\n",
       "  {'dev_ideal': 0.028,\n",
       "   'dev_ideal_target': 0.04,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_angle_d'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_angle_deg'},\n",
       "  {'dev_ideal': 0.034,\n",
       "   'dev_ideal_target': 0.05,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_planar_d'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_hb_or_metal_coord'},\n",
       "  {'dev_ideal': 1.944,\n",
       "   'dev_ideal_target': 2.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_mcbond_it'},\n",
       "  {'dev_ideal': 2.417,\n",
       "   'dev_ideal_target': 3.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_mcangle_it'},\n",
       "  {'dev_ideal': 2.613,\n",
       "   'dev_ideal_target': 2.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_scbond_it'},\n",
       "  {'dev_ideal': 3.738,\n",
       "   'dev_ideal_target': 3.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_scangle_it'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_plane_restr'},\n",
       "  {'dev_ideal': 0.091,\n",
       "   'dev_ideal_target': 0.15,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_chiral_restr'},\n",
       "  {'dev_ideal': 0.164,\n",
       "   'dev_ideal_target': 0.3,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_singtor_nbd'},\n",
       "  {'dev_ideal': 0.274,\n",
       "   'dev_ideal_target': 0.3,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_multtor_nbd'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_xhyhbond_nbd'},\n",
       "  {'dev_ideal': 0.112,\n",
       "   'dev_ideal_target': 0.3,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_xyhbond_nbd'},\n",
       "  {'dev_ideal': 3.0,\n",
       "   'dev_ideal_target': 7.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_planar_tor'},\n",
       "  {'dev_ideal': 13.9,\n",
       "   'dev_ideal_target': 15.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_staggered_tor'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_orthonormal_tor'},\n",
       "  {'dev_ideal': 32.5,\n",
       "   'dev_ideal_target': 20.0,\n",
       "   'pdbx_refine_id': 'X-RAY DIFFRACTION',\n",
       "   'type': 'p_transverse_tor'},\n",
       "  {'pdbx_refine_id': 'X-RAY DIFFRACTION', 'type': 'p_special_tor'}],\n",
       " 'reflns': [{'number_obs': 49661,\n",
       "   'observed_criterion_sigma_i': 0.0,\n",
       "   'pdbx_rmerge_iobs': 0.053,\n",
       "   'pdbx_diffrn_id': ['1'],\n",
       "   'pdbx_net_iover_sigma_i': 15.0,\n",
       "   'pdbx_ordinal': 1,\n",
       "   'pdbx_redundancy': 2.4,\n",
       "   'percent_possible_obs': 95.8,\n",
       "   'dresolution_high': 1.8,\n",
       "   'dresolution_low': 15.0}],\n",
       " 'reflns_shell': [{'rmerge_iobs': 0.162,\n",
       "   'mean_iover_sig_iobs': 4.0,\n",
       "   'pdbx_ordinal': 1,\n",
       "   'percent_possible_all': 76.1,\n",
       "   'dres_high': 1.8,\n",
       "   'dres_low': 1.86}],\n",
       " 'software': [{'classification': 'refinement',\n",
       "   'name': 'REFMAC',\n",
       "   'pdbx_ordinal': 1},\n",
       "  {'classification': 'data reduction', 'name': 'DENZO', 'pdbx_ordinal': 2},\n",
       "  {'classification': 'data scaling', 'name': 'SCALEPACK', 'pdbx_ordinal': 3}],\n",
       " 'struct': {'pdbx_descriptor': 'HEMOGLOBIN, PROTOPORPHYRIN IX CONTAINING FE',\n",
       "  'title': 'DEOXY HUMAN HEMOGLOBIN'},\n",
       " 'struct_keywords': {'pdbx_keywords': 'OXYGEN TRANSPORT',\n",
       "  'text': 'OXYGEN TRANSPORT, HEME, RESPIRATORY PROTEIN, ERYTHROCYTE'},\n",
       " 'symmetry': {'int_tables_number': 4, 'space_group_name_hm': 'P 1 21 1'},\n",
       " 'rcsb_id': '1A3N'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Obtaining PDB File information\n",
    "import pypdb # for performing searches with the RCSB PDB\n",
    "\n",
    "# describe the pdb file using pypdb before downloading \n",
    "pypdb.describe_pdb('1A3N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f8237f09-18b5-47c6-9d73-c4258d2857ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1A3N.pdb\n"
     ]
    }
   ],
   "source": [
    "# Download PDB file into current directory\n",
    "fileName = downloadPDB('1A3N')\n",
    "print(fileName)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d49cb9d4-5836-4034-a493-b648359fe403",
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy import zeros\n",
    "\n",
    "# finds the center of mass (centroid) of a structure\n",
    "\n",
    "ATOMIC_NUMS = {'H':1.0, 'C':12.0, 'N':14.0, 'O':16.0, 'P':31.0, 'S':32.0}\n",
    "\n",
    "def getCenterOfMass(structure):\n",
    "    \n",
    "    centerOfMass = zeros(3, float)\n",
    "    totalMass = 0.0\n",
    "    \n",
    "    for chain in structure.chains:\n",
    "        for residue in chain.residues:\n",
    "            for atom in residue.atoms:\n",
    "                mass = ATOMIC_NUMS.get(atom.element, 12.0)\n",
    "                centerOfMass += mass * atom.coords\n",
    "                totalMass += mass\n",
    "                \n",
    "    centerOfMass /= totalMass\n",
    "    \n",
    "    return centerOfMass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f571747d-229b-4bce-8451-9f7498f5b842",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18.88902365  0.32533332  1.24039978]\n"
     ]
    }
   ],
   "source": [
    "# contains Structure Class from book examples\n",
    "# used to create a structure from a pdb file \n",
    "from Modelling import getStructuresFromFile, Structure, Chain, Residue, Atom\n",
    "\n",
    "strucObjs = getStructuresFromFile('examples/1A12.pdb')\n",
    "\n",
    "structure = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "\n",
    "print(getCenterOfMass(structure))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d87e0e57-6e4b-4632-b819-ad8c439e2f6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy import array, dot\n",
    "from Maths import getRotationMatrix\n",
    "from math import pi\n",
    "\n",
    "# rotate structure along xyz with given angle \n",
    "def rotateStructure(structure, axis=(1,0,0), angle=0):\n",
    "    \n",
    "    rMatrix = array(getRotationMatrix(axis, angle))\n",
    "    \n",
    "    for chain in structure.chains:\n",
    "        for residue in chain.residues:\n",
    "            for atom in residue.atoms:\n",
    "                newCoords = dot(rMatrix, atom.coords)\n",
    "                atom.coords = newCoords\n",
    "# since structure is rotated internally, nothing needs to be returned\n",
    "# structure object itself is altered \n",
    "rotateStructure(structure, (0,1,0), pi/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8af08344-1184-4ac2-bd42-e931d35f49c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from numpy import array, dot\n",
    "\n",
    "def getAtomCoords(structure):\n",
    "    \n",
    "    coords = []\n",
    "    atoms = []\n",
    "    \n",
    "    for chain in structure.chains:\n",
    "        for residue in chain.residues:\n",
    "            for atom in residue.atoms:\n",
    "                coords.append(atom.coords)\n",
    "                atoms.append(atom)\n",
    "                \n",
    "    return atoms, array(coords)         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b701ecb2-e179-4d9e-aa60-5515f2e32c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# roatates structure like rotateStructure()\n",
    "# innards are more simple, avoids using loops \n",
    "# uses one 2d numpy array for all the coordinates \n",
    "\n",
    "from math import pi\n",
    "\n",
    "def rotateStructureNumPy(structure, axis=array([1,0,0]), angle=0):\n",
    "    \n",
    "    rMatrix = array(getRotationMatrix(axis, angle))\n",
    "    \n",
    "    atoms, coords = getAtomCoords(structure)\n",
    "    \n",
    "    coords = dot(coords, rMatrix.T) #transpose \n",
    "    \n",
    "    for index, atom in enumerate(atoms):\n",
    "        atom.coords = list(coords[index])\n",
    "        \n",
    "rotateStructureNumPy(structure, (0,1,0), pi/2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cf4a5f84-c1a9-4b42-8292-1c85c05b12dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Cannot find license file.\n",
      " The license files (or license server system network addresses) attempted are \n",
      "listed below.  Use LM_LICENSE_FILE to use a different license file,\n",
      " or contact your software provider for a license file.\n",
      "Feature:       PYMOL_MAIN\n",
      "Filename:      /Library/Application Support/Schrodinger/licenses\n",
      "License path:  /Library/Application Support/Schrodinger/licenses:\n",
      "FlexNet Licensing error:-1,359.  System Error: 2 \"No such file or directory\"\n"
     ]
    }
   ],
   "source": [
    "# affine transformation example\n",
    "# includes transformation matrix\n",
    "# transformation matrix may be used to stretch, reflect, shrink, enlarge and rotate the structure\n",
    "\n",
    "from numpy import identity # used for identity matrix\n",
    "\n",
    "# default for the translation is a vector of zeros; so no movement occurs.\n",
    "def affineTransformStructure(structure, transform=identity(3),\n",
    "                                  translate=(0,0,0)):\n",
    "    atoms, coords = getAtomCoords(structure)\n",
    "    \n",
    "    coords = dot(coords, transform.T)\n",
    "    \n",
    "    coords = coords + translate\n",
    "    \n",
    "    for index, atom in enumerate(atoms):\n",
    "        atom.coords = coords[index]\n",
    "        \n",
    "# transformation examples:\n",
    "\n",
    "# array to mirror the structure\n",
    "mirrorTransform = array([[-1,0,0], [0,-1,0], [0,0,-1]])\n",
    "\n",
    "# array to move 10units along x and y axis \n",
    "translate = array([10.0, 10.0, 0.0])\n",
    "\n",
    "# call function and apply arrays to structure for transformation\n",
    "affineTransformStructure(structure, mirrorTransform, translate)\n",
    "\n",
    "# write the transformation of the structure to a new pdb file \n",
    "from Structures import writeStructureToFile\n",
    "writeStructureToFile(structure, 'testTransform.pdb')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7dbd48d9-7cc1-49fb-a539-d05264e409d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "THR OG1\n",
      "LYS C\n",
      "LYS O\n",
      "ASP N\n",
      "ASP CA\n",
      "ASP C\n",
      "ASP O\n",
      "ASP CB\n",
      "ASP CG\n",
      "ASP OD1\n",
      "ASP OD2\n"
     ]
    }
   ],
   "source": [
    "# calculating atomic distances and torsion angles (dihedral angles)\n",
    "\n",
    "from math import sqrt\n",
    "\n",
    "# basic example:\n",
    "#deltas = atomA.coords - atomB.coords\n",
    "#sumSquares = dot(deltas, deltas)\n",
    "#distance = sqrt(sumSquares)\n",
    "\n",
    "def findCloseAtoms(structure, xyz, limit=5.0):\n",
    "    \n",
    "    closeAtoms = []\n",
    "    xyz = array(xyz)\n",
    "    atoms, coords = getAtomCoords(structure)\n",
    "    limit2 = limit * limit\n",
    "    \n",
    "    deltas = coords - xyz\n",
    "    squares = deltas * deltas\n",
    "    sumSquares = squares.sum(axis=1)\n",
    "    \n",
    "    boolArray = sumSquares < limit2\n",
    "    indices = boolArray.nonzero()[0] # array for first and only axis\n",
    "    \n",
    "    closeAtoms = [atoms[i] for i in indices]\n",
    "    \n",
    "    return closeAtoms\n",
    "\n",
    "atoms = findCloseAtoms(structure, (18.89, 0.36, 1.24), 5.0)\n",
    "for atom in atoms:\n",
    "    print(atom.residue.code, atom.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c87a922e-aab2-4b31-851e-ea964a9e94d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate the twist angle between atoms\n",
    "# imagined as a group of four atoms in a rough rectangle\n",
    "# connected by three bonds so that three edges are connected and one edge is not.\n",
    "# angle being measured is the amount of twist between the first and last bonds, around the central bond\n",
    "\n",
    "from Maths import calcTorsionAngle\n",
    "from math import degrees\n",
    "\n",
    "def getPhiPsi(residue, inDegrees=True):\n",
    "    \n",
    "    phi = None\n",
    "    psi = None\n",
    "    \n",
    "    # gets list of all residues in input residue’s chain\n",
    "    chain = residue.chain\n",
    "    residues = chain.residues\n",
    "    \n",
    "    atomN = residue.getAtom('N')\n",
    "    atomCa = residue.getAtom('CA')\n",
    "    atomC = residue.getAtom('C')\n",
    "    \n",
    "    coordsN = atomN.coords\n",
    "    coordsCa = atomCa.coords\n",
    "    coordsC = atomC.coords\n",
    "    \n",
    "    index = residues.index(residue)\n",
    "    \n",
    "    # positional index is defined to represent the location of current residue in polypeptide chain\n",
    "    if index > 0:\n",
    "        residuePrev = residues[index-1]\n",
    "        \n",
    "        atomC0 = residuePrev.getAtom('C')\n",
    "        coordsC0 = atomC0.coords\n",
    "        \n",
    "        phi = calcTorsionAngle(coordsC0, coordsN, coordsCa, coordsC)\n",
    "        \n",
    "        if inDegrees:\n",
    "            phi = degrees(phi)\n",
    "        \n",
    "    if index < (len(residues)-1):\n",
    "        residueNext = residues[index+1]\n",
    "        \n",
    "        atomN2 = residueNext.getAtom('N')\n",
    "        coordsN2 = atomN2.coords\n",
    "\n",
    "        psi = calcTorsionAngle(coordsN, coordsCa, coordsC, coordsN2)\n",
    "        \n",
    "        if inDegrees:\n",
    "            psi = degrees(psi)\n",
    "    \n",
    "    return phi, psi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c8e17a8f-032e-4c2a-8de0-0cdcd6996500",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX8AAAD4CAYAAAAEhuazAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAx5ElEQVR4nO2df5Ac5Xnnv8+MWjCLDbOyF1saEBIcXoU9Ia21h5WokooU24utSKwFGHx27Kq4QnJl351kbuuWwEUrBx970dlQyZ19hetctgOGxQivJcsO2BaJ61TIZHW7QqwtHQKB0EiBjaVxsHaQZmef+2O6Vz09/XOme/rX86na2pmemZ5nZt5+3ud9fr3EzBAEQRDSRSZsAQRBEIT2I8pfEAQhhYjyFwRBSCGi/AVBEFKIKH9BEIQUsiBsAdzy7ne/m5ctWxa2GIIgBMyRf3oLlepcw3Elm8GK974TpZkKiqUy5nSZihkiFPI55DsUz+9ndj6r9/aL0kwFJ8/OQP+OBOCqzo6mPoMdBw8e/Gdm7jIej43yX7ZsGcbHx30959hEETufPopTpTKW5HMY7O/GQG8h9HMJQppZPrQXZmqYAIyPbMS6kX2YLZUbHn9PPof9Qxs8v5/V+cze2y9W73gG7y1XGo6/I6dgfPuHfXsfACCi18yOp9btMzZRxD1PHUaxVAYDKJbKuOepwxibKIZ6LkFIO0vyOdvjRQtFXSyVsXxoL9aN7PN07Z1yUPx2MjVLyUTx2x0PgsQo/7GJItaN7HP94+98+ijKlWrdsXKlip1PH/X83n6eSxDSiP76nbkwCyVDdY/nlCwG+7sxNlEEWZwDQFPGl5Ni1947aSRC+TdjeVvN9m6sALevKZbKniYkQUgjxuv37EwFICCfU0AACvkcHtiyEgO9Bex8+qipS8hIuVLF8O4pV+8/2N+NnJKtO6ZNMPr39pNOG79+u3RFbHz+dthZ3lY/2pJ8znT52MzyzupchItLVG1CAiCxAEHQYXb9VqqMyy5ZgEnV/62tDKxcPmaUyhXc8F9+hIULsvh1uYJ8hwJm4NflSl1cTrse2xmz276pB4NPHkKl2jiVtUtXJEL5N2PFD/Z3456nDtcNumaXd2bnIqDBQnGakCRoLKQF/Vi3suS161dbGRgnCDfMVOYwU6llDp2duehPNypY/STQDvQTjtmE5qQr/CARyr8ZK97P2d74Q2aJULVIGzNOSNpFUCyV6yYMWSkIYeO3MWI11q3Qrl+zlYEfeFWwfn8f2oRjld3UjAvaC4lQ/s1a8X7O9tp5nCwU/YRktGi8rhQEISiMY7NVY8RprBvRX79BKkE35x6bKGLHninblUMr+OmC9kIiAr4DvQU8sGUlCvlcQ4ConThZKFoMQAvouLFogp79BcEMvzPY3Frvxut3bKKIDNnl97SGk4LVJi294tfwK6PPLODcjgyjRFj+gL9WfLPYBaPMXDpuLoagZ39BMMPPbDi3rysYirQ0xWvmQnXjNnLCjYJ1mrT8MM7CCDgDCVL+YaPlH1sNSDOXjl1sAEhufrEQffx2RVidT8NsrFsp3iwRPvGBq7HrYNH0cSVDeMelC1Caqc/w0W6XyhVkiVCuVHH3E4ewdXQSBQuF66Tc/TLOwjBeRfn7hNv8Yz1VZihZMk33IqpNEFtHJ7F1dBKdHQo23rgYe184Pb8EzecUDG/uCX3FIyQPP7PhrM6nGUteFe8cM+4fWIm+axY1JFmYnUsfqL0ip9Rdc5rxZeXDt5u04m6cifL3iWaWf50dCn7z9mzdsQwAyhCqc/UTwtmZCh45cKLuWKlcwRdGJwFIRpDgL367ItyeT6+oMxYrY83admMtGwPNdu0TzBIszCYtIBmGlyh/n3Ba1hrJKVkwAxWDkp8DgDn3a4g5AMO7p2I9CIVo4rcrwul8RkVtpvi9Wtte00SNRlxY/vh24Eu2DxF9g4jeJKIXdceGiahIRJPq30d1j91DRMeI6CgR9fshQ9jYlYgbyRLh1jUF35o4lcoVLBvai94vPiMtJITYMrx7ylZRd3YonrP4vK7IzXz4A70F7B/agOMjG7F/aEMiFD/gX6rnNwHcbHL8QWZerf79EACI6AYAdwLoUV/zVSLKmrw2Vpilm35y7dKGCQEALlUyGH3+dd9lODtTweCTh2QCEGLH2ETR0RjqWLjAs+L1EpCNuw/fK764fZj5Z0S0zOXTbwHwODOfB3CciI4BuAnAc37IEiZmy9q+axZhePdU3cA+d8H/akWNSpWlMEyIHW7y5fXtHty6Yax89kasgs5JJugir88T0QuqW6hTPVYAoDd7T6rHGiCiu4honIjGp6enAxY1GAZ6C7jsEvdzrFW3PyJAcflrSWGYEDfc9tT32sFXW5HbddHU6gvSpPiBYJX/1wBcB2A1gNMAvqweN3OFm0Y4mflhZu5j5r6uroZdyGKDW2VcyOewfVOPabXfgx9fjdnGne1MkcIwIW647anfbOXx2xXzi0fJUKpcPXoCU/7M/AYzV5l5DsDXUXPtADVL/2rdU68CcCooOaKAG2WsDW67VhVuzqNk0zuYhfjitqd+M5XHdhk/Lu2pRBJYqicRLWbm0+rdjwHQMoF2A/gOEX0FwBIA1wN4Pig5ooCT37GzQ8H2TRdzhq1S4tav6MKjB05YFpN1KBn81y03pm756gZplx1t3KZUNlN5bDcxVOcYO/akM1XaF+VPRI8B+H0A7yaikwC2A/h9IlqNmkvnVQB/CgDMPEVETwD4BYBZAJ9j5uAioBFAG1h3P3HINHfZTRbD2EQRuw4WbauIGYTx186IkjPgd4dKIRjc1BU0U3nsVINj1rQtDfji9mHmTzDzYmZWmPkqZv7fzPxHzLySmW9k5s26VQCY+UvMfB0zdzPzj/yQIeoM9BYw57LHvxluilXKlSoePXBCNpI3IHssJ4dmOvgO9nc37AksSIVvW2mlWZbboLHsCdCI3x0qhXDxWnk80Fto6MevJ5+zzgRKMono5x8XWunb3UoGT9qVnNV3J1lR6aFk49oZ3tzTRkmigyj/NtLKpjNmE4cRq4Vt2pVcWJtlCNHB6hro7FBitSrWNrJfPrR3flOoZhG3T5tptlmWWTbE+hVdePbIdN19Y49zUXLJbs4luMMqULx9U3ysfr8TF4htNhOJEn19fTw+Ph62GJFHUhoFwZy4XxvrRvaZxgyNO6AZIaKDzNxnPC6Wf8KIwnaWQPwvNCF5ROXaaBa/ExfE5y/4jtf+K4IgOON34oIof8F3vObV+xnEEoSk4nfigrh9BN/xsjyV6ltBcIffiQui/AXf8VLMZrdKEOUvCPX4GbcQt4/gO16Wp1J9KwjhIMpf8B0vxWxSfSsI4SBuHyEQ3C5Pm+nSKAhC64jyF0JFqm8FIRxE+QuhE/fiG8EaKfaLLqL8BUEIBEnjjTYS8BUEIRBkE51oI8pfEIRAkDTeaCPKXxCEQJA03mgjyl8QhECQTXSijQR8BUEIBEnjjTai/AVBCAxJ440uvrh9iOgbRPQmEb2oO7aIiH5MRC+p/zt1j91DRMeI6CgR9fshgyAIguAev3z+3wRws+HYEICfMvP1AH6q3gcR3QDgTgA96mu+SkT2O5MLgiAIvuKL24eZf0ZEywyHbwHw++rtbwH4ewD/WT3+ODOfB3CciI4BuAnAc37IIghC+5AK3vgSZLbPe5j5NACo/69UjxcAvK573kn1WANEdBcRjRPR+PT0dICiCoLgFdmuM96EkepJJsfY7InM/DAz9zFzX1dXV8BiCYLgBangjTdBKv83iGgxAKj/31SPnwRwte55VwE4FaAcgiAEgFTwxpsglf9uAJ9Rb38GwPd1x+8kokuIaDmA6wE8H6AcgiAEgFTwxhu/Uj0fQy1g201EJ4noswBGAHyIiF4C8CH1Pph5CsATAH4B4O8AfI6Zq+ZnFgQhqkgFb7zxK9vnExYP/YHF878E4Et+vLcgCOEgFbzxRip8Bd+R9L/0IBW88UWUv+ArsoGHIMQDUf6Cr9il/4nyF5JIXFe6ovwFX5H0PyFNxHmlK/38BV+R9D8hTcS50E2Uv+Arkv4nJImxiSLWjezD8qG9WDeyr6F1RZxXuuL2EXxF0v+EpODGpbMkn0PRRNHHYaUryl/wHUn/E+KEVcDWTfLCYH933QQBxGelK8pfEITUYmfdu3HpxHmlK8pfEITUYmfdu3XpxHWlKwFfQUgpTsHMNGBn3Sc9eUEsf0FIIXHOT/cTO+s+zi4dN4jyF4QUIpXYNZwCtnF16bhBlL8gpBAzaxeIR366nyTdurdDlL8gpIyxiSII5nunxiE/3W+SbN3bIQFfQUgZO58+aqr4CUhMMFNwRpS/IKQMK9cOI13B3rQjyl8QUoaVa6eQQpdPmhGfvyAkCDe95VtpSRDX3vVCI6L8BSEhuM3d127v2DOFszMVAMAlC5ydAGbn3zY6ia2jkyjIRBA7RPkLQkLwmrv/dmVu/napXKmbKDQLv1gqgwhgswgxLmYMFUtlbB2dxI49U9i+qUcmgRgQuPInolcBvAWgCmCWmfuIaBGAUQDLALwK4OPMfDZoWQQhyXjpLW81UWxVLXl9KqiV4jfj7EwllZXCcaRdAd/1zLyamfvU+0MAfsrM1wP4qXpfEIQW8LKLmlMxlwd930BcdrJKO2Fl+9wC4Fvq7W8BGAhJDkFIDGaNyJQs4dz52YbmbUEXc6WtUjiOtEP5M4BniOggEd2lHnsPM58GAPX/lWYvJKK7iGiciManp6fbIKogxJeB3gIe2LIShXwOBKCzQwG45s9nXAwAj00UAy/mSmOlcNxoh/Jfx8zvB/ARAJ8jot9z+0JmfpiZ+5i5r6urKzgJBSEhDPQWsH9oA46PbETHwgWozNU7cPQB4HxOCUSGJLU9TjKBK39mPqX+fxPA9wDcBOANIloMAOr/N4OWQxCSgts+/E4B4D9ctdh32Qr5HB7YslKCvTEgUOVPRJcR0Tu12wA+DOBFALsBfEZ92mcAfD9IOQQhKWi59sVSucGVY8QuADw2UcSug/5u3qJkSHL9Y0TQlv97APwfIjoE4HkAe5n57wCMAPgQEb0E4EPqfUEQHLDL5TditxOV2XlapTLHGN491fTrZWex9hJonj8zvwJglcnxXwH4gyDfWxCSiJdcfrte9dtGJwORr1SuYGyi6Mr6ryskQ316aVp3FmsnUuErCDHC7abiGla96q3O4wdWSlvfF+iKnIJzF2ZRqdZUvlldgT44LT2F/Ee6egpCjPBrU3Gz8/hFuVLF3U8cqnPbGGMVpXJlXvHbcapU9hTn0Lhv7DCuu+eHWDa0F9fd80PcN3bYh0+WLMTyF4QY4ee2g5csyPju99eoMmPwyUMAajI3G2PIEGGriYvKblWw7F057H/5TJ0sjxw4AQC4f2Blcx8ogYjyF4SY0eq2g8bunEFRqTJ27JnCQG+h6Yrfqk1jIf2qQN9p1Mqd9djPX0ffNYvEfaQiyl8QUkYQmT5WnJ2pBYAzRLaKvBmW5HOePkuV2VXL67QgPn9BSBnt7rtzz1OHfVf8WpzD62dxmyabBkT5C0LKaHffHb9XGfoqYj8+S1qb0InyF4SUEWSmj18oWYKSobpjBOBTa5di/9CGeTeNl89itUdxWpvQic9fEFKGMWMo36GgNFNpqYe/nxCAO/7N1QCARw+cuLipDIBdB4s4Pv0bHHjlLKrMyBJh7bWdePVX1oFeAPPbTDa7d3ESIfbZFxcUfX19PD4+HrYYgpBIxiaK2DY6GZkJIKdkcamSmd9j2Inrr7wMJ8++bepiyinZeTdRGovFiOigbiOtecTyFwQBQGu7d/lNuVL1FCt46c1zlo/duuZiamyrabJJQnz+gpBytFz5pLLrYFGaxJkgyl8QUk478/7DIM3pnHaI20cQUk4aUh3dfsY0xQRE+QtCyrHq8GlssxxrCFg2tBcAkM8pGN7cY9p11O8K4ChPJuL2EYSUY9Up9JNrl4Ykkf/okxpL5Qq2jk42dPq02ihnx57mNqhpphtpOxHLP+aYWRaA+66PUbZMhPZg1yn02SPTgfX9D5tHD5xA3zWL5j+/lWtI60/k9bqw23UtCteY5PnHmLGJIgafPFTXFz2bIWRQ21JPQ5/nbHy9WXdHq2WxkD7a1QE0LPI5BZddsgCnSmXb5nOFfA77hzZ4MpaWD+01dZsRgOMjG/37EA5Y5fmL2yfG7Ngz1bAhRnWO6xQ/YJ3tYJXlUSpXIrU8FcJjoLeAB7asRCGfAwHo7FBAjq+KD6VyZd4t47Z9tFs3jlXbiKi0kxC3T4xxW/0ImC9p7TIgorQ8FcLFWBg1NlHEvd87jHMXkrkaMMOqfbTddRJ0O4lWXbai/FOCmbXhtI+rfnKQ2ICgYZwMer/4jCdDJG5oCttq03srI8rPXdeM+JGZFJrbh4huJqKjRHSMiIbCkiPO5HOKq+dZWRtOHRG1CSPqWQtCuGzf1AMlG54z6PorL/Psiirkc+jsML9+OjuUeTeXm/bRdm6cgd4C9g9twPGRjXXdSFvFbhXillCUPxFlAfxPAB8BcAOATxDRDWHIEmeGN/c0tL1VMoRPrV1qOnjNuFQxHwL6CcOPgSYkl4HeAnbetspSmQbNS2+ew5XvXNhw3GpCyBJhsL8b2zf1mKa4bt/UY6qwrVJiw+gKarXa8FKwF5blfxOAY8z8CjNfAPA4gFtCkiW2DPQWsPP2VfN9yrNEqMwxnj0yjcH+bltrQ7Pm9ct17WIxThh+DDQh+fzm7VnT44V8zvUqtVneeOtCw7HfuW6R6cpW284RQF0w28lQMga/C/kcbl1T25x++dBerBvZ17bVsB/B5LB8/gUAr+vunwTwAeOTiOguAHcBwNKlySk4aRY7v7ud/8/sdWbWPONiSpv+Pa1S4KKStSCEz/DuqYYsM6DmmtRSJNudMjp16i3cuqaARw6caHhMW7l6dcXo4x1BVAS7xY9gcljK32xF1jBymPlhAA8DtTz/oIWKMnYDzckto68FKJbKDbUBejRrfmyiiOHdUyiVzQN5ad4EQ2jEapxoxwd6Cxh/7YypIg5SpmePTFs+7mblajSclr0rN7+RjBntypLzI5gclvI/CeBq3f2rAJwKSZZYYKfg7dwyZrUAVoofqFnzTlZalsh2eSwIZtgp4qCwy2ZjAOtG9tUpTb2yvyKn4NyF2TrDyU21c7vcoa3uTRCW8v9HANcT0XIARQB3Avi3IckSC+wUvFXKJsNbLYCSJUuXkJ45ZlH8Qh2dHYrpWNMHgf1Sih1KBudn2bYoS8OpOZ1+BT3+2pm6bSOtVjNO2LlDo5QyHYryZ+ZZIvo8gKcBZAF8g5mb656UEqwU/JJ8DutXdNUN2mapqhaOm4tU65CoR7vQClIHkDq2b+ppcCcqWcL2TT3z953qStyQU7LYsqaAHxw67Uo5u7kmtOZtfuxjbOcODTNGYEZoef7M/ENmfh8zX8fMXwpLjrhglWa2fkUXdh0s+tJ6dw61wJ1TINfqvbTjxVIZg989JHUAKUJL99Rnwuy8bVWdUnOqK7FDn10z+vzrTVvlVpz1aQN7O3eolev27ifCuVakwjcmWAV4/N6FqVSuYHhzT8uZGZU5xvDuKbH+U4STD1o/hr2sAPQZaOtG9plmFUWBQj5n+/mtVtT61NN2Xi+i/GOE2cVlVXLe6vsY/Z/N4Ld1JsQf/Ri+b+ywY/aP0Y0SRDA1p2RdGTpZIqy9thPPvXIGxvnHTfabndsrjF5a0tUz5vida68F6J49Mp2cXZyESHL/wErb4q/ODqXBjeLXeM/nlDpXkhM5JYsvf3wVHv2T38YrD2zEQ3esdl0cpuHk9mp30WQilf/YRBHrRva1veouDFrxo5pxdqaCdSP7fNnAI6xyfyEauLkOf22zOny7MtdwbLC/u6GlSTOcn53Dg3esxmB/Nx77+euOzzcq94HeAgb7u+et+bufOIRlDvpGqxDOkrn87S6aTJzbJ2oR9aDxy0Wjx6+dm/SZHkK6cHsdenWFaLf1BYgdSgaVObatXzE79/DuKZyfnXNMGTXz5Rs/n3YOJ31jVpEPhFM0mTjLP41NyKxcNFYWRjvIKZlETraCO9xeh4P93bYdQc1cIQO9BUxu/zBeHdmIV0c2ovOySzwpfo1SueLo6zcqZW01s3V00vK1TvrGrEdQGEWTibP809iEzOqzzTFbFt8ETbkyh/vGDuP+gZVtf28hfNxehwO9Bds2Im5cIUHtMUyoV+ReVthO+qbV6lw/SJzlH/Wt04LA7jOH6Xp55MAJ3Dd2OLT3F8LDy3Vo5/d34woJaoVrrFt5xINrNQ76JnHKP0o9t9uF3WcO27p45MCJRAfcBXO8XIdWijKfU1yNXzdtHlrFa21BaeZC5Md94pR/VPxp7STqn3nr6CSWDe3F6h3PRP6CEPzBy5i0miiGN7tbtRYiaGWfu1DFttHJSK98idswa/pBX18fj4+Phy1GLFm94xnHgqt8TsH52bnA+60rGcLO21dFZmISokErDc/GJorYNjoZ2bqUzg4F2zf1hDbmieggM/c1HBfln3zGJor4whOTDVWJGjkliwe21AKzVu1s/cS4YYwgtIqbauEw0a6xMCYAK+WfuGwfoRFtwO3YMzWf+WPVgVM/OI3W2PoVXXj2yHTL2RVJzrwSwuH+gZXY+8LppjPblCwBXO/bd2oH7YUw2jc4IZZ/iml2qT02UcTgdw813WBLLH8hCNxuFakZPMaxDzQ2TvRz9zECcHxkoy/n8vS+YvkLelqphNYe39pEUzklQ4nOvBLCw9g1NEMwdXWeO1/baN7MADGO/YFe8z2AmyFDhOVDe0PfxGVenlDfXQgNqwrMraOTrvohDfQWPGdZ5HOKBHuFQNF67uSUrGWMq1Su4J6nDrvOPPMrm6jKDMZFQyvszDdR/inFzu/udnC6aSqXU7J46I7VeOiO1bjskgXY5nJyEYRmcbPHhZeWL4P93fC7jCwKLWdE+acUpwpEN4PTLJf7U2uXNuR2A7VGVsVSOVKWj5BM3CYUuH3eQG+h6cCvXWfbsBMfxOefUgb7ux2DY24Gp5seJetG9lk2+RIXkOA3bvcK9tKCoWBzziyRaZVxPqegZJN9FHYLCLH8U4rearfCr8GZxmZ7Qni4dUcO9ne73vvD7JxKhtDZoaDK3OAWIgB/uGqx5TVEcNe3KEhE+aeYgd4C9g9twEN3rA60H1Iam+0J4RGEO9J4znxOAQjzdQVGu58B7DpYxPoVXQ3XFgH45Nqloa96A8vzJ6JhAH8CYFo99OfM/EP1sXsAfBZAFcB/YOannc4nef7B0kp5vZtzm21eoa94DPL9BX9x+q3sHo/S72y1Y52bOhS3u91Z1RS08zOHlef/IDP/d4MgNwC4E0APgCUAfkJE72PmYJvKCLYE2V9cn39tpRDStPtanHH6reweBxCp37kVd6SXoHIUevebEUbA9xYAjzPzeQDHiegYgJsAPBeCLEKbsLsA7HZ9iuJFk2acfiurx4d3T+Gtt2cbAqNh/s5WgWE37sgggsrtJmif/+eJ6AUi+gYRdarHCgD0OyafVI81QER3EdE4EY1PT0+bPUVIABIQjg9Wv0mxVMbyob2WCrFUrlj23Q/rd25l7w8vQeWo0pLyJ6KfENGLJn+3APgagOsArAZwGsCXtZeZnMp0VDDzw8zcx8x9XV1drYgqRBgJCMcHu9+k2ehhWL9zK/tguA0qR3nl2pLbh5k/6OZ5RPR1AD9Q754EcLXu4asAnGpFDiHemNUcRN1qSipOAVk39SFeULLh9npqxR8fVV++WwLz+RPRYmY+rd79GIAX1du7AXyHiL6CWsD3egDPByWHEH2cAsJCe3ATeDf+VnbWPqFm1c9cmLVutRyDpsJRylDykyBTPf8WNZcPA3gVwJ9qkwER3QvgjwHMAtjKzD9yOp+kegpCsDST+ujmNU6tlqPc4jsJacpWqZ6BBXyZ+Y+YeSUz38jMm3WrADDzl5j5OmbudqP4BUEInmYC726Cppp/3Ov7RgG77Cbg4uQQx75VUuErCAKA5gLvboOmdi3AoxzYd5oQnSaHKCON3YREE/UleZRoNvDuNvAZx8C+Uy1AnNOUxfIXEkucl+Rh0ErqYxTOHwRObq04pynLHr5CYmmld4sgaDj1KnIKCIeN7OErpI44L8nTRNRdc3ZurTinKYvyFxJLK71bBH9w0wE0Ss3emiGuxV7i8xcSSyu9W5KC281Kgnpvp5hLnLNl4o4ofyGxxDHA6CdhB7zdKHZxzYWHuH2ERBPXJbkfhN0q241iF9dceIjyF4SACSugGbZVbaXY8x0K1o3sw6lSGVfkFChZQqV6Meswba65sBC3jyAESJiul7Bz0E03Pc8SfvP27Pz3USpXAAY6O5RUuubCJHHKP8wAlyAYCTOgGXbA2yzmctnCBajM1dcWVeYYZ2cqyHcoOHd+FttGJ+XabQOJcvskIW1M8Jewc8jDdL34nYPezHdpjLksH9pr+Vx922e5doMnUco/7ACXEC2iYAyEHdD0K+Dt13fpdu9bQK7doEmU2yfsAJcQLaKQQx6268Uv/Pou3ex9q6dYKosbNyASZfmHbWUJ0SIKxkCcy//1+PVd6r8PNysAAuafJ64gf0mU8o9jy1ghOKJiDLh1vYQdn7Cjle/S7HPtH9rguMMXoXGXR3EF+Uei3D5pr+gU6omTyyXsalwnmv0u7T6X8Xrt7FDQoVxUSVb9hsWN6w/S0llINFG2pvXEof10M9+ll8/ltBKwe20YxGVsSUtnIZXEpb1DFOITTjTzXVrJXyyV561/DbOgshEC5lcbYSrfKGSStUqi3D6CEFfCrsYNCjv5jW4tNxMdo6Zcg3aTORWLRiGTrFVaUv5EdDsRTRHRHBH1GR67h4iOEdFRIurXHV9DRIfVx/6aiKgVGQQhCcQpPuGF9Su6YHWB65Xl2EQRGReqQNsE3kr5Du+eakleTRaniSUOKzUnWrX8XwSwBcDP9AeJ6AYAdwLoAXAzgK8SkTayvwbgLgDXq383tyiDIMSeOCUruG2hMjZRxK6DRcvALVBTlpqyrTrEH/WToZWSLZUr8/I02+rFjVWfhJVaSz5/Zv4lAJgY77cAeJyZzwM4TkTHANxERK8CuJyZn1Nf920AAwB+1IocgpAE4hCf8OLrduPDX5LPuXpeweDTt6sU1pR0sz55N1Z9EtLKg/L5FwC8rrt/Uj1WUG8bj5tCRHcR0TgRjU9PTwciqCAI7nFjFWsWt5sirsH+bkdXCQHYP7ShTmnbKdlTpbKlnFstmsbpVwlW7ie9VR+nlZoVjpY/Ef0EwHtNHrqXmb9v9TKTY2xz3BRmfhjAw0At1dNBVEEQAsbJKnabrgkA+ZyCgd6CY7WvmStloLeAHXum6prB6Z9vdz7jKsAos5n7ycyqj8NKzQ5H5c/MH2zivCcBXK27fxWAU+rxq0yOC4IQA67IKbUe/AYYtZz+c+dnXSn+nJLF8OYeAOYuFP3zzKz8+8YOo2Si+IFakPmxn79uG0MoV6q4+4lD2DY6iQyR6XOzRJhjjnQOfysElee/G8B3iOgrAJagFth9npmrRPQWEa0F8HMAnwbwNwHJIAiCj4xNFHHuwqzl4267dWaJ6lwkxn4/WVUZG/38GveNHcYjB05Ynt/uMT2awreaJOaYcXxko6tzxZGWlD8RfQw15d0FYC8RTTJzPzNPEdETAH4BYBbA55hZm9b/HYBvAsihFuiVYK8gxICdTx+t226xGXJK1tQ3rneh6Iu3tFiC/vmP/fx1tINWM3eiXgEs7R0EQXDF8qG9tmmbbujsULDxxsV49si0qVI0ixkYJ4xlNhvC+IX2nkBzHVmtYh/5nILhzT1tnQSkvYMgCC3hZSMWK87OVOrcMsbgq12Wzs6nj2Kwv3veLRQk7196BQBg8LuH5redLJbKGPzuoXlZzdCsfavvqVSuRKYNhLR38IjsESykFS8bsXjZsKVcqWLHnlplrl3apzZRrL220/W5jWSJcNlCZ9n2v3wGg9+dNN1v2KqKWF8ZbEdU2kCI5e+BJDRzEoRmGegtYPy1M3j0wAlb90+WCLeuKeDZI9OuVwpnZ2qVuU6ri3KlivHXSvjU2qWOGT1mVJlxYXbOdK8AI5U58+OlcgXLhvY2BKbdFKtpRKENhCh/D8gewUJSaDYY+eyRaUelWWXGIwdOgAjoUDKYsdKiBoZ3T6FSdX7u+dk5HJ/+DV5+4KMAgNU7njFNP7XCaM03izbxaEagW8UPRKMNhCh/DyShmVPUiXqGRBIwW8FuG53E1tFJFPI5rF/RZRmQ9TLWmeFa8QPwpMD3v3ymLYFft5QrVdexiKi0gRDl74GobAuYVMSt1h7MVrCayiqWyqYB2fHXzriy+qOEkgHmOPjgsEaVGTkl25CppLnAombQiPL3QBKaOUWZqLjVkr768LpSLVeqrgunokRt0dG+6SpDtSyh/S+fmT/2/qVX4P6BlW2TwQuS7eOBJDRzijJRcKtFfS9dP5CVajDMMeoUP1C7f9/Y4ZAkskcsf4/EvZlTlImCWy0qq48gGezvxuCTh1qu1vWTfE7BuQuzkZLJLx49cKJp6z/IVahY/kJkiMJuVlFYfbSFCOnYDAHDm3uw87ZV6OxQwhbHdxhoauUY9CpULH8hMugbfIXlb4/C6sMrXqzDsYkito1ORkn349IFmbomb06N2+LI3U/YVwabEfQqVJS/ECnCdqvFLajvJUNqbKKIL0RM8QON6aA/OHQ6JEmCo8pclzXlZqIOehUqyl8QdERh9eEFK+tQ61Wvl3949xTcZ92Hh5d8/zhRrlTrqqP1EzXQOOaCXoWK8hcEA2GvPrxgZQUaq0+B6CrVfE7BfWOHm2rXEDeMn07ra/R2Za5h9XbrmgJGn3+9riJZyZBvq1AJ+ApCjHFjBUalkZgVXe9ciEcOnEi84rfi7EzFdPX2g0OnGze+Nd9euClE+QtCBHHbPdZtp81TpbKrbpbthgC89Oa5sMWIJKVypSH1tVJl3yZycfsIQsTwEsQ1xihAtZ46RjoWZnFhNnoe/3Ta+vW46TCqRwK+gpBQvKb46WMUVh0uZy5URdFGFK+/y5J8zpfiL3H7CELEaCXF79cWQV1R/Mkgp2SxfkVXQ/HXttFJz20kRPkLQsSwCuK6Ce5aPSdLPkYKhVDo7FDwwJaVePbItGlX1kcPnPBU/SvKXxAiRittLqxe+4kPXO1pa0UherytFsNZrQAZ8BQMFuUvCCFjzOwB0HT3WKvOs/cPrJw/LsQTLe5jtwL0EgwmbiG3lohuBzAM4LcA3MTM4+rxZQB+CUCbhg4w85+pj60B8E0AOQA/BPAf2YUQfX19PD4+3rSsghBFjJk9QM1Sb1er8LGJIu5+4lBqc+zjBgF48I7Vlv2ZCvkc9g9tqH8N0UFm7jM+t1XL/0UAWwD8zOSxl5l5tfr3Z7rjXwNwF4Dr1b+bW5RBEGKLXWZPOxjoLWAuxoo/Q7UK4bSwJJ/DQG8Bn1y7tKHey2sPqpaUPzP/kpldj1IiWgzgcmZ+TrX2vw1goBUZBCHORKGFdJQ7ljpx+aUKJrd/OGwx2sa587MYmyji/oGVePCO1S1tLBWkz385EU0Q0T8Q0e+qxwoATuqec1I9ZgoR3UVE40Q0Pj09HaCoghAOrWT2+IXbKuF2kCXCqyMb8ZCq2JzQUlvTEssolSsYfPIQVu94BttGJwHU3ED7hzb4n+dPRD8hohdN/m6xedlpAEuZuRfAFwB8h4guh3lnCss1JzM/zMx9zNzX1dXlJKogxI4obGCjBYlb2UjFr0TST3zg6nmZ9g9tcFTq2iQZ1ZbbQVCpMkrlSssbvDhW+DLzB72elJnPAziv3j5IRC8DeB9qlv5VuqdeBeCU1/MLQlKISgtprUpYqxwtlsqu2w7klGxD3MINOSWDtytz8++RUzLou2ZR3XMG+7stg5sEYP2KLqwb2Ze8ndY80OwGL4G0dyCiLgBnmLlKRNeiFth9hZnPENFbRLQWwM8BfBrA3wQhQxAEuZ+mkF6i1EJaL4t+IrDj1jUF23bM+ZwCIqA0U6m7boyZTuXKXEMPo4HeAsZfO1PXBx+oKf7fuW4Rdh0sNjXxJI1mJr9WUz0/hpry7gJQAjDJzP1EdCuALwKYBVAFsJ2Z96iv6cPFVM8fAfj3cUj1DDslTxDCYt3IPssJoLNDqetFb0TJEHbevmr+GnHTt98sXdHM8HKamAjAwgUZnI9gQzu/MfvONKxSPVuy/Jn5ewC+Z3J8F4BdFq8ZB/CvW3nfMAh6P01BiCp2ViUzbC3vytzFFsR//tQLDVs2mlEslbF6xzMNqwWjctMCnkYIwPGRjfP33a5g4kIGqNuRrdkYkVT4uiQKKXmA+z7vguAXVplH+Zxi2UhOjxaUdKP4NUrlCs7O2Ac13WZKuQ0exwa1tqHZFE8NUf4uiUJKnuZ60nfz018UMjEIQWCVkTS8ucf1+G/VL29W+LZ+RZdtoZPxeli/ogtKNv4N7uYYOD8713SKp4b083fJYH+3qc+/nSlmTtWgbjcACRIJiicPp4wk43URFPpV9thEEbsOFhuygEg9YrYhzq6DRSzIUMPuWHHED5ezKH+XBJmS51Zh2rmeohCT8LIDlRAvzDKStHHbrmwb/SrD6n1n1IyhS5WM6fWQJFp1OYvy90AQKXleFOaSfM40aGV1XDtfu4jCBCS0B7PstyAxrrLtFF+5Uo2Mos9QzU0TBK26nMXnHzI79ky5buxlVw1qtVlHOzfxiEpQXAgOzY++dXSyZQVrHMtKhtDZUQtk5nNK3YbzlyyoV1XNKr7ODiXwVhZZovlg7OWXtt50LqdkAqkCF+UfImMTRZydMc+WMFOYA70F3LqmMK/Qs0S4dU1tNWKVN93OVr1RCIoLwaFPOPAD474DO29fhYm/+DCOj2zE8OaeOou5VK7UJTc49SPK5xqVfE7JYvumHjywZWVgnUBzShZf/vgqHB/ZiP1DG1xlQ9mhZAgPbLmx6f0d7BC3T4jYte01U5hakEtT6FVm7DpYRN81i1CwcP20M70tCkFxITj89O8vzJKtG9UpuUF7nKhWa6BHy0TSnmcWSxvoLeC+scMNlcOtohljGnYuWT1mrTTyOQXDm3vqZPYTUf4t0Gpmi507xExh2l0QUVC8xqD4FWpZ/7bRyXkZxfcfL/Rj3E8l+Ve3rbJ93OraKJbKGPzuIVTUZQFzzTp+x6ULGtpHAPYK8/6Blei7ZlHdNbx+RReePTI9P35LHi33Z4/Udx8e7O/GF0YnYVbhkFMyeGDLjXU9ldqZJSfKv0n8yGyxsgryOcVzto9TNpK+yjFLhCozCgEMMn2DMMn8iTdjE0UMPnnIVWqkfiw5VdR2dpiPbz12FnPFEEGtzDE6Fi7AxF947+vvlMTR+8VnLF2zZhiv0YHeAnbsmTI9x6LLLqmbpNp9XYjPv0n82IHJrnjGDCefulbJqPkb9Ypf76vV3EattIN1IuwdqoTW2bFnylHx55QsHjIUG2nj8KE7Vlv63Z3wusdAUEkF2zf1NBSGKVmyjBmYXaMlD3G9diLKv0n8yGyx2mzbygJotve7na82KIUsmT/xx87idTNevY5vs9e6JaikgoHeAnbetqo+MH3bKgxv7nF9LUY1EaKlrp7thIimAbzWxrd8N4B/tnpQ6Vq2krILFhqPc3X2QmX61cNBCZXJXb4o+45FBcouWMjV2QvV35wpzpX/5YzuKQ1yL3zvv1rjdN4L/3TsoJ9yNvn92H7nESaucgM2stuNG7/HixVW46g682tkO66o3WGem/2X6dcM10HguLgW55+34PKua0CUmZc9d3k7Zb6GmRt2w4qN8m83RDRu1gY16sRVbiC+ssdVbiC+ssdVbiA6sovbRxAEIYWI8hcEQUghovyteThsAZokrnID8ZU9rnID8ZU9rnIDEZFdfP6CIAgpRCx/QRCEFCLKXxAEIYWkXvkT0e1ENEVEc0TUpzu+jIjKRDSp/v0v3WNriOgwER0jor8mamPfZAe51cfuUWU7SkT9UZLbCBENE1FR9z1/VPeY6eeIEkR0syrfMSIaClseO4joVfX3nySicfXYIiL6MRG9pP7vDFtOACCibxDRm0T0ou6YpaxRGSsWckdzjDNzqv8A/BaAbgB/D6BPd3wZgBctXvM8gN9GrdDxRwA+EiG5bwBwCMAlAJYDeBlANipym3yOYQD/yeS45eeIyh+ArCrXtQAWqvLeELZcNvK+CuDdhmN/BWBIvT0E4L+FLacqy+8BeL/+GrSSNUpjxULuSI7x1Fv+zPxLZnbd34CIFgO4nJmf49ov+G0AA0HJZ4WN3LcAeJyZzzPzcQDHANwUFbk9YPo5QpbJyE0AjjHzK8x8AcDjqMkdJ24B8C319rcQkTHBzD8DYKx+tZI1MmPFQm4rQpU79crfgeVENEFE/0BEv6seKwA4qXvOSfVYVCgAeF13X5MvynJ/noheUJfM2lLe6nNEiTjIqIcBPENEB4noLvXYe5j5NACo/68MTTpnrGSNw+8QuTGeipbORPQTAO81eeheZv6+xctOA1jKzL8iojUAxoioBzWXiZFA8mWblNtKvrbJbcTucwD4GoC/VGX5SwBfBvDHCFFeD8RBRj3rmPkUEV0J4MdEdCRsgXwi6r9DJMd4KpQ/M3+widecB3BevX2QiF4G8D7UZuerdE+9CsApP+Q0kcGz3KjJd7XuviZf2+Q24vZzENHXAfxAvWv1OaJEHGSch5lPqf/fJKLvoeZieIOIFjPzadU1+GaoQtpjJWukfwdmfkO7HaUxLm4fC4ioi4iy6u1rAVwP4BV1ufkWEa1Vs2U+DcDKCg+D3QDuJKJLiGg5anI/H1W51YtY42MAtCwJ08/Rbvkc+EcA1xPRciJaCOBO1OSOHER0GRG9U7sN4MOofde7AXxGfdpnEIExYYOVrJEeK5Ed42FExKP0p/4YJ1Gz8t8A8LR6/FYAU6hF4/8vgE261/SpP+DLAP4H1ErpKMitPnavKttR6DJ6oiC3yef4WwCHAbyA2sWw2OlzROkPwEcB/D9VznvDlsdGzmvVsXxIHdf3qsffBeCnAF5S/y8KW1ZVrsdQc71W1HH+WTtZozJWLOSO5BiX9g6CIAgpRNw+giAIKUSUvyAIQgoR5S8IgpBCRPkLgiCkEFH+giAIKUSUvyAIQgoR5S8IgpBC/j9kH9etFTxPeAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# twist angle example and plot:\n",
    "\n",
    "from matplotlib import pyplot\n",
    "\n",
    "phiList = []\n",
    "psiList = []\n",
    "\n",
    "for chain in structure.chains:\n",
    "    for residue in chain.residues[1:-1]:\n",
    "        phi, psi = getPhiPsi(residue)\n",
    "        phiList.append(phi)\n",
    "        psiList.append(psi)\n",
    "\n",
    "pyplot.scatter(phiList, psiList) # creates a scatter plot\n",
    "pyplot.axis([-180,180,-180,180]) # set bounds\n",
    "pyplot.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c618f8ea-8c1a-421f-b77d-b918bc1cd8ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# structure subsets\n",
    "\n",
    "from Modelling import Structure, Chain, Residue, Atom\n",
    "\n",
    "# function taking in structure and three other arguments\n",
    "# arguments used to declare chain, residue, or atom \n",
    "# creates a new structure out of the subset section\n",
    "\n",
    "def filterSubStructure(structure, chainCodes=None,\n",
    "                       residueIds=None, atomNames=None):\n",
    "    \n",
    "    name = structure.name + '_filter'\n",
    "    \n",
    "    if chainCodes:\n",
    "        name += ' ' + ','.join(chainCodes)\n",
    "        chainCodes = set(chainCodes)\n",
    "        \n",
    "    if residueIds:\n",
    "        name += ' ' + ','.join([str(x) for x in residueIds])\n",
    "        residueIds = set(residueIds)\n",
    "    \n",
    "    if atomNames:\n",
    "        name += ' ' + ','.join(atomNames)\n",
    "        atomNames = set(atomNames)\n",
    "    \n",
    "    conf = structure.conformation\n",
    "    pdbId = structure.pdbId\n",
    "    \n",
    "    # intialize a new structureObj for the filtered struc\n",
    "    # keeping the same name and pdbId for the obj to indicate origin of data\n",
    "    filterStruc = Structure(name=name, conformation=conf, pdbId=pdbId)\n",
    "    \n",
    "    # loop through all chains, residues, atoms of inputs\n",
    "    # crearting lists of items to copy and use when creating filtered structure\n",
    "    for chain in structure.chains:\n",
    "        if chainCodes and (chain.code not in chainCodes):\n",
    "            continue \n",
    "        \n",
    "        includeResidues = [] \n",
    "        \n",
    "        for residue in chain.residues:\n",
    "            if residueIds and (residue.seqId not in residueIds):\n",
    "                continue\n",
    "            \n",
    "            includeAtoms = []\n",
    "            \n",
    "            for atom in residue.atoms:\n",
    "                if atomNames and (atom.name not in atomNames):\n",
    "                    continue\n",
    "                \n",
    "                includeAtoms.append(atom)\n",
    "            \n",
    "            if includeAtoms:\n",
    "                includeResidues.append((residue, includeAtoms))\n",
    "        \n",
    "        if includeResidues:\n",
    "            filterChain = Chain(filterStruc, chain.code, chain.molType)\n",
    "            \n",
    "            for residue, atoms in includeResidues:\n",
    "                filterResidue = Residue(filterChain, residue.seqId, residue.code)\n",
    "                \n",
    "                for atom in atoms:\n",
    "                    coords = array(atom.coords)\n",
    "                    Atom(filterResidue, name=atom.name, coords=coords, element=None)\n",
    "                    \n",
    "    return filterStruc\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "91e9cb12-7df5-4570-acb9-fbd7558e5ede",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nchainCodes = set(['A'])\\nresidueIds = None  #No residue filter: all of them\\natomNames = set(['N', 'CA', 'C']) # Heavy backbone atoms (not H)\\n\\nchain_A_backbone = filterSubStructure(structure, chainCodes, residueIds, atomNames)\\n\""
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# filterStruc test example\n",
    "'''\n",
    "chainCodes = set(['A'])\n",
    "residueIds = None  #No residue filter: all of them\n",
    "atomNames = set(['N', 'CA', 'C']) # Heavy backbone atoms (not H)\n",
    "\n",
    "chain_A_backbone = filterSubStructure(structure, chainCodes, residueIds, atomNames)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ea8974df-545c-46cc-b03e-430e739aab6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def to copy a structure, uses the filterfunction\n",
    "# passes no additional params other than structure\n",
    "# results in a complete copy of the structure\n",
    "\n",
    "def copyStructure(structure):\n",
    "    \n",
    "    return filterSubStructure(structure, None, None, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ccc1a472-b400-4688-8b17-90729500f232",
   "metadata": {},
   "outputs": [],
   "source": [
    "# centering coordinates\n",
    "\n",
    "from numpy import zeros, ones, cross, sqrt, linalg, exp, identity\n",
    "\n",
    "def centerCoords(coords, weights):\n",
    "    \n",
    "    wCoords = coords.transpose() * weights\n",
    "    \n",
    "    xyzTotals = wCoords.sum(axis=1)\n",
    "    \n",
    "    center = xyzTotals/sum(weights)\n",
    "    \n",
    "    coords -= center # moves all points so center is at 0,0,0\n",
    "    \n",
    "    return coords, center"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4d7a0417-de87-4f64-a7d6-5fb7fedcc98b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aligning Coordinates\n",
    "\n",
    "def alignCoords(coordsA, coordsB, weights=None):\n",
    "    \n",
    "    n = len(coordsA)\n",
    "    if weights is None:\n",
    "        weights = ones(n)\n",
    "        \n",
    "    rMat = dot(coordsB.transpose()*weights, coordsA)\n",
    "    \n",
    "    rMat1, scales, rMat2 = linalg.svd(rMat)\n",
    "    \n",
    "    sign = linalg.det(rMat) * linalg.det(rMat2)\n",
    "    \n",
    "    if sign < 0:\n",
    "        rMat1[:,2] *= -1\n",
    "        \n",
    "    rotation = dot(rMat1, rMat2)\n",
    "    \n",
    "    coordsB = dot(coordsB, rotation)\n",
    "    \n",
    "    return rotation, coordsB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "91de2366-8e74-42e9-8569-e65e2fefdd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Root-mean-square deviation\n",
    "\n",
    "def calcRmsds(refCoords, allCoords, weights):\n",
    "    \n",
    "    rmsds = []\n",
    "    totalWeight = sum(weights)\n",
    "    totalSquares = zeros(refCoords.shape)\n",
    "    \n",
    "    for coords in allCoords:\n",
    "        delta = coords-refCoords\n",
    "        squares = delta * delta\n",
    "        totalSquares += squares\n",
    "        sumSquares = weights*squares.sum(axis=1)\n",
    "        rmsds.append(sqrt(sum(sumSquares)/totalWeight))\n",
    "        \n",
    "        nStruct = len(allCoords)\n",
    "        atomRmsds = sqrt(totalSquares.sum(axis=1)/nStruct)\n",
    "        \n",
    "    return rmsds, atomRmsds\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "673e1a2d-c64a-42f8-9341-d4c086eef6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aligning a structure ensemble \n",
    "\n",
    "# superimpose more than two coordinate arrays\n",
    "# necessary to superpose all the conformations of a whole structure ensemble\n",
    "\n",
    "def superimposeCoords(allCoords, weights, threshold=5.0):\n",
    "    \n",
    "    \n",
    "    nStruct = len(allCoords)\n",
    "    \n",
    "    refCoords = allCoords[0]\n",
    "    meanCoords = zeros(refCoords.shape)\n",
    "    rotations = []\n",
    "    \n",
    "    for index, coords in enumerate(allCoords):\n",
    "        if index == 0:\n",
    "            rotation = identity(3)\n",
    "        else:\n",
    "            rotation, coords = alignCoords(refCoords, coords, weights)\n",
    "            allCoords[index] = coords # Update to aligned\n",
    "            \n",
    "        rotations.append(rotation)\n",
    "        meanCoords += coords\n",
    "    \n",
    "    meanCoords /= nStruct\n",
    "    \n",
    "    rmsds, atomRmsds = calcRmsds(meanCoords, allCoords, weights)\n",
    "    \n",
    "    bestRmsd = min(rmsds)\n",
    "    bestIndex = rmsds.index(bestRmsd) # closest to mean\n",
    "    bestCoords = allCoords[bestIndex]\n",
    "    \n",
    "    weightScale = atomRmsds/threshold #scale the RMDS value by a threshold value\n",
    "    #new weights are calculated as the negative exponent of the scale values squared\n",
    "    weights *= exp(-weightScale*weightScale)\n",
    "    \n",
    "    # atoms with small RMSD values will have the largest weights and as the variance \n",
    "    #   increases the weighting will diminish exponentially.\n",
    "    \n",
    "    meanCoords = bestCoords.copy()\n",
    "    \n",
    "    # don’t have to add best coords to the average, they are already in required array, \n",
    "    #        and because it is the reference there is no need to do the superimposition\n",
    "    \n",
    "    for index, coords in enumerate(allCoords):\n",
    "        if index != bestIndex:\n",
    "            rotation, coords = alignCoords(bestCoords, coords, weights)\n",
    "            rotations[index] = rotation \n",
    "            allCoords[index] = coords # Update to aligned \n",
    "            meanCoords += coords\n",
    "            \n",
    "    # summation of coordinate arrays is divided by number of strucs \n",
    "    #     to get revised avg (mean) coord set\n",
    "    \n",
    "    meanCoords /= nStruct\n",
    "    \n",
    "    weights = ones(len(weights))\n",
    "    rmsds, atomRmsds = calcRmsds(meanCoords, allCoords, weights)\n",
    "    \n",
    "    return allCoords, rmsds, atomRmsds, rotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "53c0e353-e16c-457f-afa6-7ce2ab0eee06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# superimpose structures\n",
    "# applies the optimized superimposeCoords() to structure objects\n",
    "\n",
    "def superimposeStructures(structures): # takes in list of structure objects\n",
    "    # coords and atoms for strucobjs are extracted \n",
    "    # coords are superimposed\n",
    "    # updates coords of original atom objects, resulting in structure rotation\n",
    "    \n",
    "    weights = None\n",
    "    allCoords = []\n",
    "    \n",
    "    for structure in structures: \n",
    "        atoms, coords = getAtomCoords(structure)\n",
    "        \n",
    "        if weights is None:\n",
    "            weights = ([ATOMIC_NUMS[atom.element] for atom in atoms])\n",
    "            weights = array(weights)\n",
    "        \n",
    "        coords, center = centerCoords(coords, weights)\n",
    "        allCoords.append(coords)\n",
    "   \n",
    "    # superimpose coords\n",
    "    results = superimposeCoords(allCoords, weights)\n",
    "    allCoords, rmsds, atomRmsds, rotations = results\n",
    "    \n",
    "    # loop through strucutre objects\n",
    "    for i, structure in enumerate(structures):\n",
    "        # get list of atoms and oldCoords (original)\n",
    "        atoms, oldCoords = getAtomCoords(structure)\n",
    "        # loop through atom index (j) and atoms\n",
    "        for j, atom in enumerate(atoms):\n",
    "            # update coords of atom obj to newly moved coords\n",
    "            atom.coords = allCoords[i][j]\n",
    "    \n",
    "    return rmsds, atomRmsds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "949153b0-7875-4125-8588-a57ab3b6e329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[7.678437039667182e-15, 7.68983871186093e-15]\n",
      "\n",
      "strucObjs list example:\n",
      " [2.921344325211597, 2.3360837755497266, 2.5800238925790273, 2.368921677536993, 3.0340759139505296, 3.1599159667158374, 3.9985636965400695, 3.6476272866881705, 2.137260125371427, 3.039174809325437]\n"
     ]
    }
   ],
   "source": [
    "# testing superimposition function\n",
    "\n",
    "from Modelling import getStructuresFromFile\n",
    "\n",
    "# load file twice, into two strucObjs\n",
    "strucA = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "strucB = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "rotateStructureNumPy(strucA, (1,0,0), pi/3) # only rotate one\n",
    "\n",
    "rmsds, atomRmsds = superimposeStructures([strucA, strucB])\n",
    "\n",
    "print(rmsds)\n",
    "\n",
    "\n",
    "# can also get whole ensemble and test alignment of the strucObjs list\n",
    "\n",
    "fileName = 'examples/1UST.pdb'\n",
    "strucObjs = getStructuresFromFile(fileName)\n",
    "\n",
    "rmsds, atomRmsds = superimposeStructures(strucObjs)\n",
    "print(\"\\nstrucObjs list example:\\n\", rmsds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6f120a57-2a97-47a2-9b6b-d1d0d71332cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# homologous structure alignment\n",
    "\n",
    "# find subsection of coords that is common between both struc\n",
    "# do a seq align to find like residue subsections\n",
    "# first must convert from structures and residues to one letter codes from 3 letter codes\n",
    "\n",
    "# 3 to 1 dictionary\n",
    "THREE_LETTER_TO_ONE = {'ALA':'A','CYS':'C','ASP':'D','GLU':'E',\n",
    "                       'PHE':'F','GLY':'G','HIS':'H','ILE':'I',\n",
    "                       'LYS':'K','LEU':'L','MET':'M','ASN':'N',\n",
    "                       'PRO':'P','GLN':'Q','ARG':'R','SER':'S',\n",
    "                       'THR':'T','VAL':'V','TRP':'W','TYR':'Y',\n",
    "                       'G':'G','C':'C','A':'A','T':'T','U':'U'}\n",
    "# loop through residues and convert to single letter codes\n",
    "# use single letter string for each struc to align \n",
    "\n",
    "def getChainSequence(chain):\n",
    "    \n",
    "    letters = []\n",
    "    \n",
    "    for residue in chain.residues:\n",
    "        code = residue.code\n",
    "        letter = THREE_LETTER_TO_ONE.get(code, 'X') # passes X if single letter cannot be determined\n",
    "        \n",
    "    seq = ''.join(letters)\n",
    "    \n",
    "    return seq\n",
    "\n",
    "\n",
    "\n",
    "from Alignments import sequenceAlign, BLOSUM62\n",
    "\n",
    "\n",
    "def seqStructureBackboneAlign(chainA, chainB, \n",
    "                              atomNames=set(['CA', 'C', 'N']),\n",
    "                              simMatrix=BLOSUM62):\n",
    "   \n",
    "    structureA = chainA.structure\n",
    "    structureB = chainB.structure\n",
    "    residuesA = chainA.residues\n",
    "    residuesB = chainB.residues\n",
    "    \n",
    "    seqA = getChainSequence(chainA)\n",
    "    seqB = getChainSequence(chainB)\n",
    "    \n",
    "    score, alignA, alignB = sequenceAlign(seqA, seqB, simMatrix)\n",
    "    \n",
    "    # empty lists to contain the location of the aligned residues\n",
    "    pairedPosA = []\n",
    "    pairedPosB = []\n",
    "    # position counters as we loop through each seq\n",
    "    posA = 0\n",
    "    posB = 0\n",
    "    \n",
    "    for i in range(len(alignA)):\n",
    "        # No dashes in both seq at the same location\n",
    "        \n",
    "        if alignA[i] == '-':\n",
    "            posB += 1\n",
    "        \n",
    "        elif alignB[i] == '-':\n",
    "            posA += 1\n",
    "        \n",
    "        else:\n",
    "            pairedPosA.append(posA)\n",
    "            pairedPosB.append(posB)\n",
    "            posA += 1\n",
    "            posB += 1\n",
    "    \n",
    "    # uses paired alignment positions, residue indices, to \n",
    "    #    get correct ResidueObj and put its number (seqId) in a list\n",
    "    filterIdsA = [residuesA[p].seqId for p in pairedPosA]\n",
    "    filterIdsB = [residuesB[p].seqId for p in pairedPosB]\n",
    "    \n",
    "    # filterSubStructure() used to define new Structure objects representing only backbone atoms.\n",
    "    backboneStrucA = filterSubStructure(structureA, [chainA.code],\n",
    "                                        filterIdsA, atomNames)\n",
    "    backboneStrucB = filterSubStructure(structureB, [chainB.code],\n",
    "                                        filterIdsB, atomNames)\n",
    "    # collect list of atoms and coords for each backbone Struc\n",
    "    atomsA, coordsA = getAtomCoords(backboneStrucA)\n",
    "    atomsB, coordsB = getAtomCoords(backboneStrucB)\n",
    "    weights = ones(len(atomsA)) # weights are all equal when initialized \n",
    "    \n",
    "    # move backbone structures to the center\n",
    "    coordsA, centerA = centerCoords(coordsA, weights)\n",
    "    coordsB, centerB = centerCoords(coordsB, weights)\n",
    "    \n",
    "    # input a list of coords to superimnpose the necessary atoms\n",
    "    coords = [coordsA, coordsB]\n",
    "    c, rmsds, atomRmsds, rotations = superimposeCoords(coords, weights)\n",
    "    \n",
    "    #transform the original structures from a rotational transformation that was recorded\n",
    "    affineTransformStructure(structureA, rotations[0], -centerA)\n",
    "    affineTransformStructure(structureB, rotations[1], -centerB)\n",
    "    \n",
    "    return rmsds, atomRmsds\n",
    "\n",
    "# example \n",
    "# not working... files are different sizes, chain lengths are different\n",
    "#struc1 = getStructuresFromFile(downloadPDB('1UST'))[0]\n",
    "#struc2 = getStructuresFromFile(downloadPDB('1HST'))[0]\n",
    "#chain1 = struc1.getChain('A')\n",
    "#chain2 = struc2.getChain('A')\n",
    "#rmsds, atomRmsds = seqStructureBackboneAlign(chain1, chain2)\n",
    "#print(rmsds)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "020dffac-f238-4b0e-a470-909cf113811c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# external macromolecular structure models\n",
    "\n",
    "# examples are not working stright out of book\n",
    "\n",
    "\n",
    "\n",
    "# structures in bio python\n",
    "\n",
    "#from Bio import PDB\n",
    "\n",
    "# make parser object that can make a PDB.Structure object using the data from a file.\n",
    "#fileName = 'examples/1A12.pdb'\n",
    "#parser = PDB.PDBParser()\n",
    "#struc  = parser.get_structure('Name', fileName)\n",
    "#conformation = struc[0]\n",
    "\n",
    "#for chain in conformation:\n",
    "    #for residue in chain:\n",
    "        #atomNames = [a.name for a in residue]\n",
    "        #print(chain.id, residue.id[1], residue.resname, atomNames)\n",
    "        #caAtom = residue['CA']\n",
    "        #print(caAtom.name, caAtom.coord, caAtom.bfactor)\n",
    "        \n",
    "\n",
    "#outFileName = 'test.pdb'\n",
    "#writer = PDB.PDBIO()\n",
    "#writer.set_structure(struc)\n",
    "#writer.save(outFileName)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# structures in pymol\n",
    "\n",
    "#import pymol\n",
    "\n",
    "#pymol.finish_launching()\n",
    "\n",
    "#fileName = downloadPDB('examples/1AFO')\n",
    "#strucName = 'Glycophorin'\n",
    "#pymol.cmd.load(fileName, strucName)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f1b83f1-cc0b-4837-bb39-fccd68a605a2",
   "metadata": {},
   "source": [
    "---\n",
    "## Assignemnt:  Calculating Root-Mean-Square Deviation (RMSD) & aligning a Structure Ensemble \n",
    "---\n",
    "### The Python code for calculating the sum of the square values along the spatial axis (i.e., x2 + y2 + z2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4def31f5-40ba-4b7a-bc36-2b3d320223cf",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1329.86571288 1272.32916245 1191.10659995 ...  296.44373987  276.36671847\n",
      "  216.2503414 ]\n"
     ]
    }
   ],
   "source": [
    "# takes in structure and center of mass (xyz)\n",
    "# calls the funciton to get the coordinates of the atoms \n",
    "from Modelling import getStructuresFromFile, Structure, Chain, Residue, Atom\n",
    "\n",
    "from numpy import zeros, array, dot\n",
    "\n",
    "# finds the center of mass (centroid) of a structure\n",
    "\n",
    "ATOMIC_NUMS = {'H':1.0, 'C':12.0, 'N':14.0, 'O':16.0, 'P':31.0, 'S':32.0}\n",
    "\n",
    "def getCenterOfMass(structure):\n",
    "    \n",
    "    centerOfMass = zeros(3, float)\n",
    "    totalMass = 0.0\n",
    "    \n",
    "    for chain in structure.chains:\n",
    "        for residue in chain.residues:\n",
    "            for atom in residue.atoms:\n",
    "                mass = ATOMIC_NUMS.get(atom.element, 12.0)\n",
    "                centerOfMass += mass * atom.coords\n",
    "                totalMass += mass\n",
    "                \n",
    "    centerOfMass /= totalMass\n",
    "    \n",
    "    return centerOfMass\n",
    "\n",
    "def getAtomCoords(structure):\n",
    "    \n",
    "    coords = []\n",
    "    atoms = []\n",
    "    \n",
    "    for chain in structure.chains:\n",
    "        for residue in chain.residues:\n",
    "            for atom in residue.atoms:\n",
    "                coords.append(atom.coords)\n",
    "                atoms.append(atom)\n",
    "                \n",
    "    return atoms, array(coords) \n",
    "\n",
    "def getSumSquares(structure, xyz):\n",
    "    \n",
    "    xyz = array(xyz)\n",
    "    atoms, coords = getAtomCoords(structure)\n",
    "    \n",
    "    deltas = coords - xyz\n",
    "    squares = deltas * deltas\n",
    "    sumSquares = squares.sum(axis=1)\n",
    "    \n",
    "    return sumSquares\n",
    "\n",
    "strucObj = getStructuresFromFile('examples/1A12.pdb')\n",
    "\n",
    "structure1 = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "\n",
    "center_coords = getCenterOfMass(structure1)\n",
    "\n",
    "sumOfSquares = getSumSquares(structure1, center_coords)\n",
    "print(sumOfSquares)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78f0890-7ae3-4c5a-a1f7-c526dd472faa",
   "metadata": {},
   "source": [
    "---\n",
    "### The Python code that use \"for\" loops for getting the summed values of all and then give a single value (divided by the total weight to find the average atomic square deviation)\n",
    "\n",
    "AND\n",
    "\n",
    "### The Python code for aligning all the coordinate arrays to the reference set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d17c18fd-cdc5-48f7-adbb-6c397d5280c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[7.678437039667182e-15, 7.68983871186093e-15]\n",
      "[9.91025644e-15 9.05767819e-15 9.33644097e-15 ... 5.36824692e-15\n",
      " 5.37183335e-15 5.51994135e-15]\n"
     ]
    }
   ],
   "source": [
    "from numpy import zeros, ones, cross, sqrt, linalg, exp, identity\n",
    "from Modelling import getStructuresFromFile\n",
    "\n",
    "# Aligning Coordinates\n",
    "def alignCoords(coordsA, coordsB, weights=None):\n",
    "    \n",
    "    n = len(coordsA)\n",
    "    if weights is None:\n",
    "        weights = ones(n)\n",
    "        \n",
    "    rMat = dot(coordsB.transpose()*weights, coordsA)\n",
    "    \n",
    "    rMat1, scales, rMat2 = linalg.svd(rMat)\n",
    "    \n",
    "    sign = linalg.det(rMat) * linalg.det(rMat2)\n",
    "    \n",
    "    if sign < 0:\n",
    "        rMat1[:,2] *= -1\n",
    "        \n",
    "    rotation = dot(rMat1, rMat2)\n",
    "    \n",
    "    coordsB = dot(coordsB, rotation)\n",
    "    \n",
    "    return rotation, coordsB\n",
    "\n",
    "\n",
    "# centering coordinates\n",
    "def centerCoords(coords, weights):\n",
    "    \n",
    "    wCoords = coords.transpose() * weights\n",
    "    \n",
    "    xyzTotals = wCoords.sum(axis=1)\n",
    "    \n",
    "    center = xyzTotals/sum(weights)\n",
    "    \n",
    "    coords -= center # moves all points so center is at 0,0,0\n",
    "    \n",
    "    return coords, center\n",
    "\n",
    "\n",
    "# Root-mean-square deviation\n",
    "def calcRmsds(refCoords, allCoords, weights):\n",
    "    \n",
    "    rmsds = []\n",
    "    totalWeight = sum(weights)\n",
    "    totalSquares = zeros(refCoords.shape)\n",
    "    \n",
    "    for coords in allCoords:\n",
    "        delta = coords-refCoords\n",
    "        squares = delta * delta\n",
    "        totalSquares += squares\n",
    "        sumSquares = weights*squares.sum(axis=1)\n",
    "        rmsds.append(sqrt(sum(sumSquares)/totalWeight))\n",
    "        \n",
    "        nStruct = len(allCoords)\n",
    "        atomRmsds = sqrt(totalSquares.sum(axis=1)/nStruct)\n",
    "        \n",
    "    return rmsds, atomRmsds\n",
    "\n",
    "\n",
    "\n",
    "# Aligning a structure ensemble:\n",
    "\n",
    "# superimpose more than two coordinate arrays\n",
    "# necessary to superpose all the conformations of a whole structure ensemble\n",
    "def superimposeCoords(allCoords, weights, threshold=5.0):\n",
    "    \n",
    "    nStruct = len(allCoords)\n",
    "    \n",
    "    refCoords = allCoords[0]\n",
    "    meanCoords = zeros(refCoords.shape)\n",
    "    rotations = []\n",
    "    \n",
    "    for index, coords in enumerate(allCoords):\n",
    "        if index == 0:\n",
    "            rotation = identity(3)\n",
    "        else:\n",
    "            rotation, coords = alignCoords(refCoords, coords, weights)\n",
    "            allCoords[index] = coords # Update to aligned\n",
    "            \n",
    "        rotations.append(rotation)\n",
    "        meanCoords += coords\n",
    "    \n",
    "    meanCoords /= nStruct\n",
    "    \n",
    "    rmsds, atomRmsds = calcRmsds(meanCoords, allCoords, weights)\n",
    "    \n",
    "    bestRmsd = min(rmsds)\n",
    "    bestIndex = rmsds.index(bestRmsd) # closest to mean\n",
    "    bestCoords = allCoords[bestIndex]\n",
    "    \n",
    "    weightScale = atomRmsds/threshold #scale the RMDS value by a threshold value\n",
    "    #new weights are calculated as the negative exponent of the scale values squared\n",
    "    weights *= exp(-weightScale*weightScale)\n",
    "    \n",
    "    # atoms with small RMSD values will have the largest weights and as the variance \n",
    "    #   increases the weighting will diminish exponentially.\n",
    "    \n",
    "    meanCoords = bestCoords.copy()\n",
    "    \n",
    "    # don’t have to add best coords to the average, they are already in required array, \n",
    "    #        and because it is the reference there is no need to do the superimposition\n",
    "    \n",
    "    for index, coords in enumerate(allCoords):\n",
    "        if index != bestIndex:\n",
    "            rotation, coords = alignCoords(bestCoords, coords, weights)\n",
    "            rotations[index] = rotation \n",
    "            allCoords[index] = coords # Update to aligned \n",
    "            meanCoords += coords\n",
    "            \n",
    "    # summation of coordinate arrays is divided by number of strucs \n",
    "    #     to get revised avg (mean) coord set\n",
    "    \n",
    "    meanCoords /= nStruct\n",
    "    \n",
    "    weights = ones(len(weights))\n",
    "    rmsds, atomRmsds = calcRmsds(meanCoords, allCoords, weights)\n",
    "    \n",
    "    return allCoords, rmsds, atomRmsds, rotations\n",
    "\n",
    "# superimpose structures\n",
    "# applies the optimized superimposeCoords() to structure objects\n",
    "\n",
    "def superimposeStructures(structures): # takes in list of structure objects\n",
    "    # coords and atoms for strucobjs are extracted \n",
    "    # coords are superimposed\n",
    "    # updates coords of original atom objects, resulting in structure rotation\n",
    "    \n",
    "    weights = None\n",
    "    allCoords = []\n",
    "    \n",
    "    for structure in structures: \n",
    "        atoms, coords = getAtomCoords(structure)\n",
    "        \n",
    "        if weights is None:\n",
    "            weights = ([ATOMIC_NUMS[atom.element] for atom in atoms])\n",
    "            weights = array(weights)\n",
    "        \n",
    "        coords, center = centerCoords(coords, weights)\n",
    "        allCoords.append(coords)\n",
    "   \n",
    "    # superimpose coords\n",
    "    results = superimposeCoords(allCoords, weights)\n",
    "    allCoords, rmsds, atomRmsds, rotations = results\n",
    "    \n",
    "    # loop through strucutre objects\n",
    "    for i, structure in enumerate(structures):\n",
    "        # get list of atoms and oldCoords (original)\n",
    "        atoms, oldCoords = getAtomCoords(structure)\n",
    "        # loop through atom index (j) and atoms\n",
    "        for j, atom in enumerate(atoms):\n",
    "            # update coords of atom obj to newly moved coords\n",
    "            atom.coords = allCoords[i][j]\n",
    "    \n",
    "    return rmsds, atomRmsds\n",
    "\n",
    "# load file twice, into two strucObjs\n",
    "strucA = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "strucB = getStructuresFromFile('examples/1A12.pdb')[0]\n",
    "\n",
    "# calls function to rotate\n",
    "# only rotate one so we can measure the difference in the rmsds value\n",
    "\n",
    "rotateStructureNumPy(strucA, (1,0,0), pi/3) \n",
    "\n",
    "# aligns structures, by moving coordinates in 3d space\n",
    "rmsds, atomRmsds = superimposeStructures([strucA, strucB])\n",
    "\n",
    "print(rmsds)\n",
    "\n",
    "print(atomRmsds)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3b8cc46-5a48-4c8b-a781-2f671aca4203",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
